platform/pulp/plugin/changeset/iterator.py

```diff
@@ -0,0 +1,215 @@
+                yield (content, fetched)
+
+    def _iter(self):
+        """
+        The remote content is iterated and its model is replaced with a model
+        fetched from the DB (when found). This is batched to limit the memory
+        footprint.
+
+        Yields:
+            RemoteContent: The transformed content.
+        """
+        for batch, fetched in self._fetch():
+            for content in batch:
+                natural_key = content.model.natural_key()
+                try:
+                    model = fetched[natural_key]
+                except KeyError:
+                    pass
+                else:
+                    content.update(model)
+                yield content
+
+    def __iter__(self):
+        return iter(self._iter())
+
+
+class DownloadIterator:
+    """
+    Iterate the content artifacts and yield the appropriate download object.
+    When downloading is deferred or the artifact is already downloaded,
+    A NopDownload is yielded.
+    """
+
+    def __init__(self, content, deferred=False):
+        """
+        Args:
+            content (generator): A generator of RemoteContent.
+            deferred (bool): When true, downloading is deferred.
+        """
+        self.content = content
+        self.deferred = deferred
+
+    def _iter(self):
+        """
+        Iterate the content artifacts and yield the appropriate download object.
+        When downloading is deferred or the artifact is already downloaded,
+        A NopDownload is yielded.
+
+        Yields:
+            Download: The appropriate download object.
+        """
+        for artifact in ArtifactIterator(self.content):
+            if self.deferred or artifact.model.downloaded:
+                download = NopDownload()
+                download.attachment = artifact
+            else:
+                download = artifact.download
+            yield download
+
+    def __iter__(self):
+        return iter(self._iter())
+
+
+class ArtifactIterator:
+    """
+    Iterate the content and flatten the artifacts.
+    Ensure that at least (1) artifact is yielded for each content.
+    """
+
+    def __init__(self, content):
+        """
+        Args:
+            content (generator): A generator of RemoteContent.
+        """
+        self.content = content
+
+    def _iter(self):
+        """
+        Iterate the content and flatten the artifacts.
+        Ensure that at least (1) artifact is yielded for each content. When content
+        does not have any un-downloaded artifacts, A NopArtifact is yielded.
+
+        Yields:
+            RemoteArtifact: The flattened artifacts.
+        """
+        for content in ContentIterator(self.content):
+            if content.artifacts:
+                for artifact in content.artifacts:
+                    yield artifact
+            else:
+                artifact = RemoteArtifact(NopArtifact(), NopDownload())
+                artifact.content = content
```

Comment:
Might be nice if this could be passed to ``__init__``. It seems a little awkward to create the instance with two arguments, and then immediately add a third thing a different way.